# --- Day 9: Stream Processing ---

> A large stream blocks your path. According to the locals, it's not safe to cross the stream at the moment because it's full of garbage. You look down at the stream; rather than water, you discover that it's a stream of characters.

> You sit for a while and record part of the stream (your puzzle input). The characters represent groups - sequences that begin with { and end with }. Within a group, there are zero or more other things, separated by commas: either another group or garbage. Since groups can contain other groups, a } only closes the most-recently-opened unclosed group - that is, they are nestable. Your puzzle input represents a single, large group which itself contains many smaller ones.

> Sometimes, instead of a group, you will find garbage. Garbage begins with < and ends with >. Between those angle brackets, almost any character can appear, including { and }. Within garbage, < has no special meaning.

> In a futile attempt to clean up the garbage, some program has canceled some of the characters within it using !: inside garbage, any character that comes after ! should be ignored, including <, >, and even another !.

> You don't see any characters that deviate from these rules. Outside garbage, you only find well-formed groups, and garbage always terminates according to the rules above.

> Here are some self-contained pieces of garbage:

    <>, empty garbage.
    <random characters>, garbage containing random characters.
    <<<<>, because the extra < are ignored.
    <{!>}>, because the first > is canceled.
    <!!>, because the second ! is canceled, allowing the > to terminate the garbage.
    <!!!>>, because the second ! and the first > are canceled.
    <{o"i!a,<{i<a>, which ends at the first >.

> Here are some examples of whole streams and the number of groups they contain:

    {}, 1 group.
    {{{}}}, 3 groups.
    {{},{}}, also 3 groups.
    {{{},{},{{}}}}, 6 groups.
    {<{},{},{{}}>}, 1 group (which itself contains garbage).
    {<a>,<a>,<a>,<a>}, 1 group.
    {{<a>},{<a>},{<a>},{<a>}}, 5 groups.
    {{<!>},{<!>},{<!>},{<a>}}, 2 groups (since all but the last > are canceled).

> Your goal is to find the total score for all groups in your input. Each group is assigned a score which is one more than the score of the group that immediately contains it. (The outermost group gets a score of 1.)

    {}, score of 1.
    {{{}}}, score of 1 + 2 + 3 = 6.
    {{},{}}, score of 1 + 2 + 2 = 5.
    {{{},{},{{}}}}, score of 1 + 2 + 3 + 3 + 3 + 4 = 16.
    {<a>,<a>,<a>,<a>}, score of 1.
    {{<ab>},{<ab>},{<ab>},{<ab>}}, score of 1 + 2 + 2 + 2 + 2 = 9.
    {{<!!>},{<!!>},{<!!>},{<!!>}}, score of 1 + 2 + 2 + 2 + 2 = 9.
    {{<a!>},{<a!>},{<a!>},{<ab>}}, score of 1 + 2 = 3.
    
> What is the total score for all groups in your input?

In [96]:
def parse_input_streams(input_list, depth_count=0, search_index=0): 
    group_count = 0
    group_sum = 0    
    list_of_group_sums = []
    open_braces = 0 
    closed_braces = 0 
    garbage = False 
    next_ignored_character = -1 
    
    for index, item in enumerate(input_list):      
        if index != next_ignored_character:            
            if item == '!':
                next_ignored_character = index + 1                
            if not garbage:
                if item == '<':
                    garbage = True
                elif item == '{':                    
                    open_braces = open_braces + 1
                    depth_count = depth_count + 1                    
                elif item == '}':
                    closed_braces = closed_braces + 1
                    list_of_group_sums.append(depth_count)
                    depth_count = depth_count - 1
            else:
                if item == '>':                    
                    garbage = False                        
    
    if open_braces == closed_braces: 
        group_count = open_braces       

    group_sum = sum(list_of_group_sums)    
    
    return (group_count, group_sum)

In [97]:
TEST_INPUT1 = '{}'
TEST_INPUT1_GROUP_COUNT = 1
TEST_INPUT1_SUM = 1
TEST1_DATA = (TEST_INPUT1, TEST_INPUT1_GROUP_COUNT, TEST_INPUT1_SUM)

TEST_INPUT2 = '{{{}}}'
TEST_INPUT2_GROUP_COUNT = 3
TEST_INPUT2_SUM = 6
TEST2_DATA = (TEST_INPUT2, TEST_INPUT2_GROUP_COUNT, TEST_INPUT2_SUM)

TEST_INPUT3= '{{{}}}'
TEST_INPUT3_GROUP_COUNT = 3
TEST_INPUT3_SUM = 6
TEST3_DATA = (TEST_INPUT3, TEST_INPUT3_GROUP_COUNT, TEST_INPUT3_SUM)

TEST_INPUT4= '{{{},{},{{}}}}'
TEST_INPUT4_GROUP_COUNT = 6
TEST_INPUT4_SUM = 16
TEST4_DATA = (TEST_INPUT4, TEST_INPUT4_GROUP_COUNT, TEST_INPUT4_SUM)

TEST_INPUT5 = '{<{},{},{{}}>}'
TEST_INPUT5_GROUP_COUNT = 1
TEST_INPUT5_SUM = 1
TEST5_DATA = (TEST_INPUT5, TEST_INPUT5_GROUP_COUNT, TEST_INPUT5_SUM)

TEST_INPUT6 = '{<a>,<a>,<a>,<a>}'
TEST_INPUT6_GROUP_COUNT = 1
TEST_INPUT6_SUM = 1
TEST6_DATA = (TEST_INPUT6, TEST_INPUT6_GROUP_COUNT, TEST_INPUT6_SUM)

TEST_INPUT7 = '{{<a>},{<a>},{<a>},{<a>}}'
TEST_INPUT7_GROUP_COUNT = 5
TEST_INPUT7_SUM = 9
TEST7_DATA = (TEST_INPUT7, TEST_INPUT7_GROUP_COUNT, TEST_INPUT7_SUM)

TEST_INPUT8 = '{{<!>},{<!>},{<!>},{<a>}}'
TEST_INPUT8_GROUP_COUNT = 2
TEST_INPUT8_SUM = 3
TEST8_DATA = (TEST_INPUT8, TEST_INPUT8_GROUP_COUNT, TEST_INPUT8_SUM)


TEST_DATA = [TEST1_DATA, TEST2_DATA, TEST3_DATA, TEST4_DATA, TEST5_DATA, TEST6_DATA, TEST7_DATA, TEST8_DATA]

for test, test_data in enumerate(TEST_DATA,1):
    groups_found, score = parse_input_streams(TEST_DATA[test - 1][0])
    
    group_test_expectation = test_data[1]
    group_sum_expectation = test_data[2]
    
    print('For test {} I found {} groups with a score of {}'.format(test, 
                                                                    groups_found,
                                                                    score))
    print('For test {} the correct number of groups were found: {}'. format(test, 
                                                                            groups_found == group_test_expectation))
    print('For test {} the correct sum of groups was found: {}'. format(test,
                                                                        score == group_sum_expectation))
    print('--')

For test 1 I found 1 groups with a score of 1
For test 1 the correct number of groups were found: True
For test 1 the correct sum of groups was found: True
--
For test 2 I found 3 groups with a score of 6
For test 2 the correct number of groups were found: True
For test 2 the correct sum of groups was found: True
--
For test 3 I found 3 groups with a score of 6
For test 3 the correct number of groups were found: True
For test 3 the correct sum of groups was found: True
--
For test 4 I found 6 groups with a score of 16
For test 4 the correct number of groups were found: True
For test 4 the correct sum of groups was found: True
--
For test 5 I found 1 groups with a score of 1
For test 5 the correct number of groups were found: True
For test 5 the correct sum of groups was found: True
--
For test 6 I found 1 groups with a score of 1
For test 6 the correct number of groups were found: True
For test 6 the correct sum of groups was found: True
--
For test 7 I found 5 groups with a score of 9

In [138]:
PUZZLE_INPUT = []

with open('./assets/day9_puzzle_input.txt', 'r') as puzzle_file: 
    PUZZLE_INPUT = puzzle_file.read()

groups_found, score = parse_input_streams(PUZZLE_INPUT)

print('For star 1 I found {} groups with a score of {}'.format(groups_found,
                                                                score))
print('My puzzle answer is {}'.format(score))

For star 1 I found 1065 groups with a score of 10800
My puzzle answer is 10800


This granted me the first star. 

# --- Part Two ---

> Now, you're ready to remove the garbage.

> To prove you've removed it, you need to count all of the characters within the garbage. The leading and trailing < and > don't count, nor do any canceled characters or the ! doing the canceling.

    <>, 0 characters.
    <random characters>, 17 characters.
    <<<<>, 3 characters.
    <{!>}>, 2 characters.
    <!!>, 0 characters.
    <!!!>>, 0 characters.
    <{o"i!a,<{i<a>, 10 characters.
    
> How many non-canceled characters are within the garbage in your puzzle input?



In [245]:
def parse_input_streams_star2(input_list, depth_count=0, search_index=0): 
    group_count = 0
    group_sum = 0    
    list_of_group_sums = []
    open_braces = 0 
    closed_braces = 0 
    
    garbage = False
    garbage_changed = 0
    garbage_count_ignore = False
    garbage_characters = 0
    
    next_ignored_character = -1 
       
    for index, item in enumerate(input_list):        
        if index != next_ignored_character:                
            garbage_count_ignore = False            
            if item == '!':
                next_ignored_character = index + 1
                garbage_count_ignore = True
            if not garbage:
                if item == '<':
                    garbage = True
                    garbage_changed = index
                    garbage_count_ignore = True
                elif item == '{':                    
                    open_braces = open_braces + 1
                    depth_count = depth_count + 1                    
                elif item == '}':
                    closed_braces = closed_braces + 1
                    list_of_group_sums.append(depth_count)
                    depth_count = depth_count - 1
            else:                
                if item == '>':                    
                    garbage = False
                    garbage_changed = index
                    garbage_count_ignore = True
                    
            if garbage and not garbage_count_ignore:
                garbage_characters = garbage_characters + 1            
                                       
    if open_braces == closed_braces: 
        group_count = open_braces       

    group_sum = sum(list_of_group_sums)    
    
    return (group_count, group_sum, garbage_characters)

In [246]:
STAR2_TEST1_INPUT = '<>'
STAR2_TEST1_ANSWER = 0
STAR2_TEST1_DATA = (STAR2_TEST1_INPUT, STAR2_TEST1_ANSWER)

STAR2_TEST2_INPUT = '<random characters>'
STAR2_TEST2_ANSWER = 17
STAR2_TEST2_DATA = (STAR2_TEST2_INPUT, STAR2_TEST2_ANSWER)

STAR2_TEST3_INPUT = '<<<<>'
STAR2_TEST3_ANSWER = 3
STAR2_TEST3_DATA = (STAR2_TEST3_INPUT, STAR2_TEST3_ANSWER)

STAR2_TEST4_INPUT = '<{!>}>'
STAR2_TEST4_ANSWER = 2 
STAR2_TEST4_DATA = (STAR2_TEST4_INPUT, STAR2_TEST4_ANSWER)

STAR2_TEST5_INPUT = '<!!>'
STAR2_TEST5_ANSWER = 0
STAR2_TEST5_DATA = (STAR2_TEST5_INPUT, STAR2_TEST5_ANSWER)

STAR2_TEST6_INPUT = '<!!!>>'
STAR2_TEST6_ANSWER = 0
STAR2_TEST6_DATA = (STAR2_TEST6_INPUT, STAR2_TEST6_ANSWER)

STAR2_TEST7_INPUT = '<{o"i!a,<{i<a>'
STAR2_TEST7_ANSWER = 10
STAR2_TEST7_DATA = (STAR2_TEST7_INPUT, STAR2_TEST7_ANSWER)

STAR2_TEST8_INPUT = '<random><characters>'
STAR2_TEST8_ANSWER = 16
STAR2_TEST8_DATA = (STAR2_TEST8_INPUT, STAR2_TEST8_ANSWER)

STAR2_TEST9_INPUT = '<random!><characters>'
STAR2_TEST9_ANSWER = 17
STAR2_TEST9_DATA = (STAR2_TEST9_INPUT, STAR2_TEST9_ANSWER)

STAR2_TEST10_INPUT = '{<!!a>}'
STAR2_TEST10_ANSWER = 1
STAR2_TEST10_DATA = (STAR2_TEST10_INPUT, STAR2_TEST10_ANSWER)

STAR2_TEST_DATA = [STAR2_TEST1_DATA, STAR2_TEST2_DATA, STAR2_TEST3_DATA,
                  STAR2_TEST4_DATA, STAR2_TEST5_DATA, STAR2_TEST6_DATA,
                  STAR2_TEST7_DATA, STAR2_TEST8_DATA, STAR2_TEST9_DATA, 
                  STAR2_TEST10_DATA]

for test, test_data in enumerate(STAR2_TEST_DATA,1):
    _, _, garbage_characters = parse_input_streams_star2(STAR2_TEST_DATA[test - 1][0])
    
    garbage_characters_expectation = test_data[1]    
    
    print('For test {} with input {} I found {} garbage characters.'.format(test,
                                                                            STAR2_TEST_DATA[test - 1][0],
                                                                            garbage_characters))    
    print('For test {} the correct number of garbage characters were found: {}'. format(test,
                                                                        garbage_characters == garbage_characters_expectation))
    print('--')

For test 1 with input <> I found 0 garbage characters.
For test 1 the correct number of garbage characters were found: True
--
For test 2 with input <random characters> I found 17 garbage characters.
For test 2 the correct number of garbage characters were found: True
--
For test 3 with input <<<<> I found 3 garbage characters.
For test 3 the correct number of garbage characters were found: True
--
For test 4 with input <{!>}> I found 2 garbage characters.
For test 4 the correct number of garbage characters were found: True
--
For test 5 with input <!!> I found 0 garbage characters.
For test 5 the correct number of garbage characters were found: True
--
For test 6 with input <!!!>> I found 0 garbage characters.
For test 6 the correct number of garbage characters were found: True
--
For test 7 with input <{o"i!a,<{i<a> I found 10 garbage characters.
For test 7 the correct number of garbage characters were found: True
--
For test 8 with input <random><characters> I found 16 garbage chara

In [249]:
PUZZLE_INPUT = []

with open('./assets/day9_puzzle_input.txt', 'r') as puzzle_file: 
    PUZZLE_INPUT = puzzle_file.read()

print(len(PUZZLE_INPUT))

star2_groups, star2_group_sum, garbage_characters = parse_input_streams_star2(PUZZLE_INPUT)

print('For star 2 I found {} garbage characters'.format(garbage_characters))
print('For star 2 I found {} groups with a score of {}'.format(star2_groups,
                                                            star2_group_sum))
print('Both answers match star 1: {}'.format((star2_groups == groups_found) and (star2_group_sum == score)))
print('My star 2 puzzle answer is {}'.format(garbage_characters))

13448
For star 2 I found 4522 garbage characters
For star 2 I found 1065 groups with a score of 10800
Both answers match star 1: True
My star 2 puzzle answer is 4522


My puzzle answer is 7283
In [ ]:
This failed as too high


My final answer was 4522 and that earned the star. 